File: backend/pkg/diff/diff.go
package diff

import (
	"fmt"
	"math"
	"strings"

	"github.com/delicious/delicious/pkg/model"
)

const amountEpsilon = 1e-6

// Compare 对比 base（左/旧/百科）与 target（右/新/我的），返回结构化 Diff。
// 配料按 name 归一化匹配；步骤按 order 匹配。时间复杂度 O(n+m)。
func Compare(base, target VersionSnapshot) VersionDiffResult {
	ingDiffs := diffIngredients(base.Ingredients, target.Ingredients)
	procDiffs := diffProcessSteps(base.ProcessSteps, target.ProcessSteps)

	return VersionDiffResult{
		IngredientDiffs: ingDiffs,
		ProcessDiffs:    procDiffs,
		Summary:         buildSummary(ingDiffs, procDiffs),
	}
}

func diffIngredients(base, target []model.Ingredient) []IngredientDiff {
	baseMap := indexIngredients(base)
	targetMap := indexIngredients(target)

	seen := make(map[string]struct{}, len(baseMap)+len(targetMap))
	var result []IngredientDiff

	// 按 target 顺序输出，便于前端右栏对齐
	for _, t := range target {
		key := ingredientKey(t.Name)
		seen[key] = struct{}{}
		b, ok := baseMap[key]
		if !ok {
			tCopy := t
			result = append(result, IngredientDiff{
				Type:   TypeAdded,
				Target: &tCopy,
			})
			continue
		}
		result = append(result, compareIngredientPair(b, t))
	}

	// base 中剩余项为 REMOVED，保持 base 原顺序
	for _, b := range base {
		key := ingredientKey(b.Name)
		if _, ok := seen[key]; ok {
			continue
		}
		bCopy := b
		result = append(result, IngredientDiff{
			Type: TypeRemoved,
			Base: &bCopy,
		})
	}

	return result
}

func compareIngredientPair(base, target model.Ingredient) IngredientDiff {
	if ingredientsEqual(base, target) {
		bCopy, tCopy := base, target
		return IngredientDiff{
			Type:   TypeUnchanged,
			Base:   &bCopy,
			Target: &tCopy,
		}
	}
	delta := target.Amount - base.Amount
	bCopy, tCopy := base, target
	return IngredientDiff{
		Type:        TypeModified,
		Base:        &bCopy,
		Target:      &tCopy,
		AmountDelta: &delta,
	}
}

func ingredientsEqual(a, b model.Ingredient) bool {
	return ingredientKey(a.Name) == ingredientKey(b.Name) &&
		floatEqual(a.Amount, b.Amount) &&
		normalizeUnit(a.Unit) == normalizeUnit(b.Unit) &&
		strings.TrimSpace(a.Note) == strings.TrimSpace(b.Note)
}

func diffProcessSteps(base, target []model.ProcessStep) []ProcessStepDiff {
	baseMap := indexProcessSteps(base)
	targetMap := indexProcessSteps(target)

	seen := make(map[int]struct{}, len(baseMap)+len(targetMap))
	var result []ProcessStepDiff

	for _, t := range target {
		seen[t.Order] = struct{}{}
		b, ok := baseMap[t.Order]
		if !ok {
			tCopy := t
			result = append(result, ProcessStepDiff{
				Type:   TypeAdded,
				Order:  t.Order,
				Target: &tCopy,
			})
			continue
		}
		result = append(result, compareProcessPair(b, t))
	}

	for _, b := range base {
		if _, ok := seen[b.Order]; ok {
			continue
		}
		bCopy := b
		result = append(result, ProcessStepDiff{
			Type:  TypeRemoved,
			Order: b.Order,
			Base:  &bCopy,
		})
	}

	return result
}

func compareProcessPair(base, target model.ProcessStep) ProcessStepDiff {
	if processStepsEqual(base, target) {
		bCopy, tCopy := base, target
		return ProcessStepDiff{
			Type:   TypeUnchanged,
			Order:  base.Order,
			Base:   &bCopy,
			Target: &tCopy,
		}
	}
	bCopy, tCopy := base, target
	return ProcessStepDiff{
		Type:   TypeModified,
		Order:  base.Order,
		Base:   &bCopy,
		Target: &tCopy,
	}
}

func processStepsEqual(a, b model.ProcessStep) bool {
	if a.Order != b.Order {
		return false
	}
	if strings.TrimSpace(a.Content) != strings.TrimSpace(b.Content) {
		return false
	}
	if !intPtrEqual(a.DurationMinutes, b.DurationMinutes) {
		return false
	}
	return strPtrEqual(a.ImageURL, b.ImageURL)
}

// indexIngredients 按归一化 name 建索引；同名重复时保留首次出现。
func indexIngredients(items []model.Ingredient) map[string]model.Ingredient {
	m := make(map[string]model.Ingredient, len(items))
	for _, item := range items {
		key := ingredientKey(item.Name)
		if _, exists := m[key]; !exists {
			m[key] = item
		}
	}
	return m
}

func indexProcessSteps(items []model.ProcessStep) map[int]model.ProcessStep {
	m := make(map[int]model.ProcessStep, len(items))
	for _, item := range items {
		if _, exists := m[item.Order]; !exists {
			m[item.Order] = item
		}
	}
	return m
}

func ingredientKey(name string) string {
	return strings.ToLower(strings.TrimSpace(name))
}

func normalizeUnit(unit string) string {
	u := strings.TrimSpace(unit)
	// 常见单位归一化，减少「克」vs「g」误判为修改
	switch strings.ToLower(u) {
	case "g", "克":
		return "g"
	case "kg", "千克", "公斤":
		return "kg"
	case "ml", "毫升":
		return "ml"
	case "l", "升":
		return "l"
	default:
		return u
	}
}

func floatEqual(a, b float64) bool {
	return math.Abs(a-b) < amountEpsilon
}

func intPtrEqual(a, b *int) bool {
	if a == nil && b == nil {
		return true
	}
	if a == nil || b == nil {
		return false
	}
	return *a == *b
}

func strPtrEqual(a, b *string) bool {
	if a == nil && b == nil {
		return true
	}
	if a == nil || b == nil {
		return false
	}
	return *a == *b
}

func buildSummary(ingDiffs []IngredientDiff, procDiffs []ProcessStepDiff) string {
	var addedIng, removedIng, modifiedIng int
	var addedProc, removedProc, modifiedProc int

	for _, d := range ingDiffs {
		switch d.Type {
		case TypeAdded:
			addedIng++
		case TypeRemoved:
			removedIng++
		case TypeModified:
			modifiedIng++
		}
	}
	for _, d := range procDiffs {
		switch d.Type {
		case TypeAdded:
			addedProc++
		case TypeRemoved:
			removedProc++
		case TypeModified:
			modifiedProc++
		}
	}

	if addedIng+removedIng+modifiedIng+addedProc+removedProc+modifiedProc == 0 {
		return "两个版本完全一致"
	}

	var parts []string
	if addedIng > 0 {
		parts = append(parts, fmt.Sprintf("新增 %d 项配料", addedIng))
	}
	if removedIng > 0 {
		parts = append(parts, fmt.Sprintf("删除 %d 项配料", removedIng))
	}
	if modifiedIng > 0 {
		parts = append(parts, fmt.Sprintf("修改 %d 项配料", modifiedIng))
	}
	if addedProc > 0 {
		parts = append(parts, fmt.Sprintf("新增 %d 个步骤", addedProc))
	}
	if removedProc > 0 {
		parts = append(parts, fmt.Sprintf("删除 %d 个步骤", removedProc))
	}
	if modifiedProc > 0 {
		parts = append(parts, fmt.Sprintf("修改 %d 个步骤", modifiedProc))
	}

	return strings.Join(parts, "，")
}

// CountIngredientByType 统计配料差异数量
func CountIngredientByType(diffs []IngredientDiff, t Type) int {
	n := 0
	for _, d := range diffs {
		if d.Type == t {
			n++
		}
	}
	return n
}

// CountProcessByType 统计步骤差异数量
func CountProcessByType(diffs []ProcessStepDiff, t Type) int {
	n := 0
	for _, d := range diffs {
		if d.Type == t {
			n++
		}
	}
	return n
}

File: frontend/h5/src/utils/diff.ts
import type { Ingredient, ProcessStep } from '@/types/recipe'
import type { DiffType, IngredientDiff, ProcessStepDiff, VersionDiffResult } from '@/types/diff'

const AMOUNT_EPSILON = 1e-6

export interface VersionSnapshot {
  ingredients: Ingredient[]
  process_steps: ProcessStep[]
}

/** 客户端 Diff（与 backend/pkg/diff 算法一致，API 不可用时本地计算） */
export function compareVersions(base: VersionSnapshot, target: VersionSnapshot): VersionDiffResult {
  const ingredientDiffs = diffIngredients(base.ingredients, target.ingredients)
  const processDiffs = diffProcessSteps(base.process_steps, target.process_steps)
  return {
    ingredient_diffs: ingredientDiffs,
    process_diffs: processDiffs,
    summary: buildSummary(ingredientDiffs, processDiffs),
  }
}

function diffIngredients(base: Ingredient[], target: Ingredient[]): IngredientDiff[] {
  const baseMap = indexIngredients(base)
  const seen = new Set<string>()
  const result: IngredientDiff[] = []

  for (const t of target) {
    const key = ingredientKey(t.name)
    seen.add(key)
    const b = baseMap.get(key)
    if (!b) {
      result.push({ type: 'added', target: { ...t } })
    } else {
      result.push(compareIngredientPair(b, t))
    }
  }

  for (const b of base) {
    const key = ingredientKey(b.name)
    if (!seen.has(key)) {
      result.push({ type: 'removed', base: { ...b } })
    }
  }

  return result
}

function compareIngredientPair(base: Ingredient, target: Ingredient): IngredientDiff {
  if (ingredientsEqual(base, target)) {
    return { type: 'unchanged', base: { ...base }, target: { ...target } }
  }
  return {
    type: 'modified',
    base: { ...base },
    target: { ...target },
    amount_delta: target.amount - base.amount,
  }
}

function diffProcessSteps(base: ProcessStep[], target: ProcessStep[]): ProcessStepDiff[] {
  const baseMap = indexProcessSteps(base)
  const seen = new Set<number>()
  const result: ProcessStepDiff[] = []

  for (const t of target) {
    seen.add(t.order)
    const b = baseMap.get(t.order)
    if (!b) {
      result.push({ type: 'added', order: t.order, target: { ...t } })
    } else {
      result.push(compareProcessPair(b, t))
    }
  }

  for (const b of base) {
    if (!seen.has(b.order)) {
      result.push({ type: 'removed', order: b.order, base: { ...b } })
    }
  }

  return result
}

function compareProcessPair(base: ProcessStep, target: ProcessStep): ProcessStepDiff {
  if (processStepsEqual(base, target)) {
    return { type: 'unchanged', order: base.order, base: { ...base }, target: { ...target } }
  }
  return { type: 'modified', order: base.order, base: { ...base }, target: { ...target } }
}

function indexIngredients(items: Ingredient[]): Map<string, Ingredient> {
  const m = new Map<string, Ingredient>()
  for (const item of items) {
    const key = ingredientKey(item.name)
    if (!m.has(key)) m.set(key, item)
  }
  return m
}

function indexProcessSteps(items: ProcessStep[]): Map<number, ProcessStep> {
  const m = new Map<number, ProcessStep>()
  for (const item of items) {
    if (!m.has(item.order)) m.set(item.order, item)
  }
  return m
}

function ingredientKey(name: string): string {
  return name.trim().toLowerCase()
}

function normalizeUnit(unit: string): string {
  const u = unit.trim()
  switch (u.toLowerCase()) {
    case 'g':
    case '克':
      return 'g'
    case 'kg':
    case '千克':
    case '公斤':
      return 'kg'
    case 'ml':
    case '毫升':
      return 'ml'
    case 'l':
    case '升':
      return 'l'
    default:
      return u
  }
}

function ingredientsEqual(a: Ingredient, b: Ingredient): boolean {
  return (
    ingredientKey(a.name) === ingredientKey(b.name) &&
    floatEqual(a.amount, b.amount) &&
    normalizeUnit(a.unit) === normalizeUnit(b.unit) &&
    (a.note ?? '').trim() === (b.note ?? '').trim()
  )
}

function processStepsEqual(a: ProcessStep, b: ProcessStep): boolean {
  return (
    a.order === b.order &&
    a.content.trim() === b.content.trim() &&
    (a.duration_minutes ?? null) === (b.duration_minutes ?? null) &&
    (a.image_url ?? null) === (b.image_url ?? null)
  )
}

function floatEqual(a: number, b: number): boolean {
  return Math.abs(a - b) < AMOUNT_EPSILON
}

function buildSummary(ingDiffs: IngredientDiff[], procDiffs: ProcessStepDiff[]): string {
  let addedIng = 0,
    removedIng = 0,
    modifiedIng = 0
  let addedProc = 0,
    removedProc = 0,
    modifiedProc = 0

  for (const d of ingDiffs) {
    if (d.type === 'added') addedIng++
    else if (d.type === 'removed') removedIng++
    else if (d.type === 'modified') modifiedIng++
  }
  for (const d of procDiffs) {
    if (d.type === 'added') addedProc++
    else if (d.type === 'removed') removedProc++
    else if (d.type === 'modified') modifiedProc++
  }

  if (addedIng + removedIng + modifiedIng + addedProc + removedProc + modifiedProc === 0) {
    return '两个版本完全一致'
  }

  const parts: string[] = []
  if (addedIng) parts.push(`新增 ${addedIng} 项配料`)
  if (removedIng) parts.push(`删除 ${removedIng} 项配料`)
  if (modifiedIng) parts.push(`修改 ${modifiedIng} 项配料`)
  if (addedProc) parts.push(`新增 ${addedProc} 个步骤`)
  if (removedProc) parts.push(`删除 ${removedProc} 个步骤`)
  if (modifiedProc) parts.push(`修改 ${modifiedProc} 个步骤`)
  return parts.join('，')
}

export function diffTypeLabel(type: DiffType): string {
  switch (type) {
    case 'added':
      return '新增'
    case 'removed':
      return '删除'
    case 'modified':
      return '修改'
    default:
      return ''
  }
}

export function formatIngredient(ing?: Ingredient): string {
  if (!ing) return '—'
  const note = ing.note ? `（${ing.note}）` : ''
  return `${ing.name}  ${ing.amount}${ing.unit}${note}`
}

export function formatDelta(delta?: number): string {
  if (delta == null || delta === 0) return ''
  return delta > 0 ? `+${delta}` : `${delta}`
}
